### PR TITLE
Advanced E2E tests

### PR DESCRIPTION
**Pull request by scottrippey**
# Advanced E2E Tests
This branch exists to implement + showcase some advanced E2E testing techniques, including:
- Mocking an entire backend via MSW
- Writing E2E tests against a mock server
- Writing E2E tests against a real production server
- Writing E2E tests that work with SSR

# Details
- Adds a mock server, powered by MSW
  - Includes mocks for GraphQL and GROQ data
- Adds a mock factory, powered by faker

# Items to complete
- [x] Fix build errors from next app (tailwindcss related)
- [x] Implement SSR spy
- [x] Improve tests to use SSR spy
- [x] Separate `e2e` and `e2e-mocks`
- [x] Investigate a better codegen for mock types (that doesn't depend on `graphql-request`)
- [x] Only allow `setMockData` when env var is set
- [x] Clean up code
- [x] Mock data storage

### CODE REVIEW

File: packages/nextjs/pages/index.tsx [hunk 1]
```diff
@@ -113,9 +119,10 @@ export const getServerSideProps: GetServerSideProps = async ({ res }) => {
   return {
     props: {
       // urqlState is a keyword here so withUrqlClient can pick it up.
-      urqlState: ssrCache.extractData(),
+      urqlState: ssrCache.extractData() as SSRData<GetProductsAndCategoriesQuery>,
     },
+    [Symbol.for("e2eData")]: query?.data,
   };
-};
+});
```
Comment (scottrippey) [inline packages/nextjs/pages/index.tsx:126]: # Open Discussion: `getPageData` versus `getServerSideProps`

To write good E2E tests against **production data**, the test needs to have access to that data.
Currently, I have implemented 2 different ways to access that data, and I'm looking for feedback on both approaches.
# TL;DR;
- `getServerSideProps` extracts the data from the `__NEXT_DATA__.props` field but `urqlState` messes with that
- `getPageData` makes an API call to retrieve the data, but it's more complex and has overhead and hacks

## `getServerSideProps`
In this approach, the E2E test visits the page, then retrieves `window.__NEXT_DATA__.props` to access the server-side props.
- It's normally quite easy to make these props type-safe (by using the return type of `[page].getServerSideProps`)
- However, MOST of the time in this app, the only server-side prop is `urqlState` as you see here.  In order to access the data, we must parse the `urqlState`, and to make the results strongly-typed, we have to do this `as SSRData<...>` cast.

## `getPageData`
The API is nearly identical, but the approach is quite different.  With this approach, we expose an HTTP endpoint, `/api/e2e-tests/page-data`.  This endpoint calls the page's actual `getServerSideProps` method, and returns the results over HTTP.
- This allows the E2E test to access the raw data
- This also allows the `getServerSideProps` call to return _different_ data.  For example, instead of only returning the `urqlState`, we can also return the raw `query.data` ... so we don't have to bother extracting the `urqlState`.  
- One caveat here is that `getServerSideProps` doesn't allow you to return non-standard fields, so I can't just add the `e2eData: query.data` to the return value.  I also don't want to add extra, unused data as a `props`, because it would get serialized.  So, I resorted to the caveat here, wrapping the value behind a `Symbol`, which doesn't throw warnings, but can still be accessed.
- Another caveat: getting the server-side data via separate API call means the data might not always be 100% in-sync


# Thoughts?

Comment (scottrippey) [reply packages/nextjs/pages/index.tsx]: I decided the best approach was `getServerSideProps`, using `__NEXT_DATA__`.  It's very clean and works everywhere.

File: packages/nextjs/mocks/msw/msw-enable.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Importing this file enables MSW.  It works both client-side and server-side.
+ */
+
+/* eslint-disable @typescript-eslint/no-var-requires */
+
+// See https://github.com/mswjs/interceptors/issues/159
+const mswUnsupported = process.version.startsWith("v18");
+if (mswUnsupported) {
+  throw new Error(`[MSW Mock Server] MSW does not support mocking 'fetch' in Node v18`);
```
Comment (kgpax) [inline packages/nextjs/mocks/msw/msw-enable.ts:13]: 😱

File: packages/nextjs/mocks/msw/msw-enable.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+async function initializeMSW() {
+  console.log("[MSW Mock Server] Starting...");
+  if (typeof window === "undefined") {
+    // 'next dev' will compile pages individually, so let's ensure we only install MSW once:
+    const isPatchedAlready = Object.getOwnPropertySymbols(require("http")).find(
+      (s) => s.toString() === "Symbol(isPatchedModule)"
+    );
```
Comment (kgpax) [inline packages/nextjs/mocks/msw/msw-enable.ts:22]: 🤔 This is interesting.  How would this present itself?  Will it be an error, or just an unnecessary memory overhead the more you use the site?  Should I look into doing this to my project as well?

File: packages/nextjs/pages/api/e2e-tests/mocks.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  if (process.env.NEXT_PUBLIC_API_MOCKING !== "enabled") {
+    res.status(404).send("Error: mocks not enabled");
+    return;
+  }
+
+  const { getMockData, setMockData } = await import("mocks/msw/db/mock-data");
+
+  if (req.method === "GET") {
+    const mockData = getMockData();
```
Comment (kgpax) [inline packages/nextjs/pages/api/e2e-tests/mocks.ts:12]: 😄 Ah - ok - this answers my question about the randomness and not knowing what values you can check for.

However, I still have a little red light flashing in my head that if a test should ever fail one time due to certain data, it might be hard to debug it since the data set that caused the issue may be hard to reproduce?

File: packages/tests-e2e/e2e-tests/categories.cy.mock.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { mock } from "mocks/factory";
+
+describe(`when I visit the "Categories" page`, () => {
+  const mockCategories = mock.categories(6);
```
Comment (kgpax) [inline packages/tests-e2e/e2e-tests/categories.cy.mock.ts:4]: 🤔 In the project I am working on, I've included my mock and real tests in the same file, with `realOnly(() => {})` and `mocksOnly(() => {})` closures.  The reason for this was that sometimes a significant portion of the tests are identical for each data source (visit this, click that, fill these fields in), and all that will sometimes change are the values being used for acting and asserting.

I'm not specifically wedded to it, and I can see the benefits and purity of your separate file approach; I just wondered whether you had encountered much duplication in both flavours of tests?

File: packages/nextjs/mocks/msw/graphql/create-graphql-handlers.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+export function createGraphqlHandlers<Sdk extends SdkHandlers>(
+  queryHandlers: Partial<Sdk>,
+  mutationHandlers: Partial<Sdk>,
+  { url = "" } = {}
+) {
+  const g = url ? graphql.link(url) : graphql;
+  return [
+    ...Object.entries(queryHandlers).map(([handlerName, handler]) => {
+      return g.query(handlerName, async (req, res, ctx) => {
+        return res(ctx.data(await handler(req.variables)));
```
[the remark this answers is not in the document]
Comment (scottrippey) [reply packages/nextjs/mocks/msw/graphql/create-graphql-handlers.ts:43]: My initial approach did include the headers as a second parameter. However, the codegen produces the `Sdk` type, which only has this `variables` signature.
I agree though, that having access to `Headers` or even the full `req` object would be ideal. I would love to make a codegen to better handle this, but obviously I could just map the Sdk myself.

File: packages/nextjs/mocks/msw/msw-enable.ts (hunk 2 above)
Comment (scottrippey) [reply packages/nextjs/mocks/msw/msw-enable.ts:13]: Yeah, right? And `urql` uses fetch by default.
Another option was to supply `node-fetch` to urql but I wasn't sure about the sanity client. I decided to just require Node 16 for now, until `msw` fixes this.

Comment (scottrippey) [reply packages/nextjs/mocks/msw/msw-enable.ts:13]: Or maybe globally replacing `fetch` with `node-fetch` if the mocks are enabled.

File: packages/nextjs/mocks/msw/msw-enable.ts (hunk 3 above)
Comment (scottrippey) [reply packages/nextjs/mocks/msw/msw-enable.ts:22]: Yes, this was some bizarre behaviour that I was seeing. I didn't know that `next dev` would behave this way. It seems like all pages are run in child processes, or something like that, so they don't share the "require" cache.
I'm not totally sure if "multiple msw" caused problems, or if they simply bypassed each other. I added this when I was debugging some performance issues, but it didn't solve it, so I don't know if it really matters. Things seemed to work either way.

File: packages/tests-e2e/e2e-tests/categories.cy.mock.ts (hunk 5 above)
Comment (scottrippey) [reply packages/tests-e2e/e2e-tests/categories.cy.mock.ts:4]: I could definitely see the benefit in your approach. The "real" and "mock" tests have a ton in common for sure.
In this codebase, I realized that basically all of my "real" tests are totally compatible with the mock server.
However, I would rather have in-file differences like you suggest

File: packages/nextjs/utils/satisfies.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Simulates behavior of TypeScript 4.9's "satisfies" operator
+ * @example
+ *   const getServerSideProps = satisfies<GetServerSideProps>()(async (context) => ({ props: { myProp1: true } });
+ *   type SSRProps = AsyncReturnType<getServerSideProps>['props']; // { myProp1: true }
+ */
+export const satisfies =
```
Comment (binoy14) [inline packages/nextjs/utils/satisfies.ts:7]: I'll create an issue to upgrade version of TS, I don't know if there should be anything that is breaking but would be good to check things before yolo upgrade

**Review by binoy14 — COMMENTED** [review]
This is really detailed and awesome thanks for doing thing, I learned a lot! I don't think I have enough expertise on some of the granular things that Kevin is pointing out so I will defer to him for the final ✅

File: packages/nextjs/utils/satisfies.ts (hunk 7 above)
Comment (binoy14) [reply packages/nextjs/utils/satisfies.ts:7]: sorry didn't realize 4.9 is not public yet, I made an [issue](https://github.com/FormidableLabs/nextjs-sanity-fe/issues/154) regardless to upgrade when it's generally available

File: packages/nextjs/pages/api/e2e-tests/mocks.ts (hunk 4 above)
Comment (scottrippey) [reply packages/nextjs/pages/api/e2e-tests/mocks.ts:12]: That's a great point. I just pushed a fix! By calling `faker.seed(0)`, we will still get random data, but at least it will be somewhat reproducible between test runs.

File: packages/nextjs/mocks/factory/index.ts [hunk 8]
```diff
@@ -0,0 +1,225 @@
+import faker from "faker";
```
[the remark this answers is not in the document]
Comment (scottrippey) [reply packages/nextjs/mocks/factory/index.ts:1]: Thanks! I feel like this is the core to writing good tests ... You need an easy way to get generic data, and you need an easy way to get specific data. This provides both!
This can be used everywhere, e2e, integration, and unit tests, and even mock development.

File: packages/nextjs/utils/satisfies.ts (hunk 7 above)
Comment (scottrippey) [reply packages/nextjs/utils/satisfies.ts:7]: Yes, good idea. I was going to upgrade to the beta, but I only needed `satisfies` for a few places, so I went with this workaround version. It will be nice to remove all that extra `<>()()` !

File: packages/nextjs/pages/api/e2e-tests/mocks.ts (hunk 4 above)
Comment (scottrippey) [reply packages/nextjs/pages/api/e2e-tests/mocks.ts:12]: And, quite by coincidence, that commit proved its worth!  The tests are now failing with: 
```
TestingLibraryElementError: Timed out retrying after 4000ms: Found multiple elements with the text: White Croissant
``` 
But more importantly -- they're failing consistently on that same error!  I ran twice, and it failed both times identically.
So this revealed a problem with my random data ... sometimes it creates duplicate category names.

**Review by gksander — APPROVED** [review]
This is... *a lot* of code. It looks great to me – nothing stood out as alarming to me, so here's a ✅ from me.

**Comment by scottrippey** [comment]
> This is... _a lot_ of code. It looks great to me – nothing stood out as alarming to me, so here's a ✅ from me.

What, 111 commits isn't that much right?

File: packages/tests-e2e/e2e-tests/categories.cy.mock.ts (hunk 5 above)
Comment (scottrippey) [reply packages/tests-e2e/e2e-tests/categories.cy.mock.ts:4]: I made this change in #155 ! Thanks for the suggestion!
